Update object degradation tutorial; rename fouling_rating to distort_extent

mabelzhang opened the pull request:
I tagged @j-herman (because you've run this in the past) or @woensug-choi (because you're looking into this now) for review. Only one is needed.

- Update object degradation tutorial to include 1. how to add more degradation methods to the Blender Python script, and 2. how to add more custom SDF tags.
  The current SDF tags are just dummy examples. @woensug-choi if you find more suitable terminology, feel free to update the example - remember to also update the Gazebo C++ plugin, see README section "To add new custom tags". Alternatively, just use whatever you like in the multibeam sonar plugin and the SDF it loads.
- Rename `fouling_rating` variable in range [0, 100] to `distort_extent`, now in range [0, 1]. This follows the discussion that we don't care about actually adhering to the navy fouling scale. It's now just an arbitrary relative scale.

## To test
Follow the mesh README to run the mesh modification script in Blender 2.92 and make sure it still works after the variable name change.

woensug-choi commented:
Is it correct that the current subdivision level from 0~4 is the number of operations to divide an edge into two with a mid vertex?

Are there methods we can define the density of the random vertexes generated by subdivision?  Something like max edge lengths? (Still discussing what we need. No need to include in this PR :))

mabelzhang commented:
> divide an edge into two with a mid vertex?

I played with it a bit on a cylinder and a cube, and it does look like what you described.

Blender's official [description](https://docs.blender.org/manual/en/latest/modeling/modifiers/generate/subdivision_surface.html) doesn't say too much about the mathematical operation. "The Subdivision Surface modifier (often shorten to “Subdiv”) is used to split the faces of a mesh into smaller faces, giving it a smooth appearance." For the "Simple" mode, which is the current mode in the script, it "Only subdivides the surfaces, without any smoothing."

For a cylinder, the top/bottom cap looks like this:

level 0
![2022-01-31-185155_3706x2049_scrot_cylinderSubdiv0](https://user-images.githubusercontent.com/7608908/151894060-42e91423-a1bf-408c-8bbc-e41ed7671843.png)

level 1
![2022-01-31-185150_3706x2049_scrot_cubeSubdiv1](https://user-images.githubusercontent.com/7608908/151894065-62ef3fbd-b937-48fd-ad76-1e7e52ffa130.png)

level 2
![2022-01-31-185146_3706x2049_scrot_cylinderSubdiv2](https://user-images.githubusercontent.com/7608908/151894072-24805c97-4d3f-4c48-a78c-167b27bb560a.png)
![2022-01-31-184413_3706x2049_scrot_cylinderSubdiv2](https://user-images.githubusercontent.com/7608908/151894103-837ede04-4550-44ae-8594-bae7d5b7dcd9.png)

mabelzhang commented:
> density of the random vertexes generated by subdivision? Something like max edge lengths?

Hmm there might be a way to script that, although I don't know if edge length can be a parameter to easily control.
The Subdiv modifier doesn't seem to expose such a parameter. There might be a different way to do it.

Reference: [Subdiv modifier options](https://docs.blender.org/api/current/bpy.types.SubsurfModifier.html)
The "Simple" mode Blender says is equivalent to the mesh operation Subdivide, which has these options https://docs.blender.org/api/current/bpy.ops.mesh.html#bpy.ops.mesh.subdivide

mabelzhang commented:
You didn't ask about this, but I guess you've already seen it and that's why you mentioned bump height:
Vertex randomization has a parameter `offset` for "Distance to offset"
https://docs.blender.org/api/current/bpy.ops.transform.html#bpy.ops.transform.vertex_random

```
bpy.ops.transform.vertex_random(offset=0.0, uniform=0.0, normal=0.0, seed=0, wait_for_input=True)
Randomize vertices

Parameters
offset (float in [-inf, inf], (optional)) – Amount, Distance to offset

uniform (float in [0, 1], (optional)) – Uniform, Increase for uniform offset distance

normal (float in [0, 1], (optional)) – Normal, Align offset direction to normals

seed (int in [0, 10000], (optional)) – Random Seed, Seed for the random number generator

wait_for_input (boolean, (optional)) – Wait for Input
```

woensug-choi commented:
Yes, the script you've made can designate offset (=bump height) magnitude (while they are also random). but cannot control the density. 

I was just asking for the density to calculate statistics for the minimum mesh edge length which would lead to average normal angles. Hmm

mabelzhang commented:
The edge length could vary a lot depending on where you are on the mesh. With a cylinder, the edges of the circles are very small, while the edges on the wall can be very long for e.g. a pipe.

Do you just want the normals? Blender has other ways to get the normals, e.g. https://docs.blender.org/api/current/bpy.types.Mesh.html

woensug-choi commented:
Don't understand how the screenshot you've shown is possible. Maybe because it went through the process `subdivide`->`offset`->`subdivide`->`offset` order? What if finish `subdivide` and `offset`? 

What I would like to have is some statistics on how much the mean normals have deviated from the original object shape. I was imagining the `vertex density` that are produced out of subdivision could provide information about `mean normal deviation`. 

Or maybe we could ditch `subdivision` function called to whole object and use a manual loop for each mesh elements and give custom subdivision density.

mabelzhang commented:
To clarify, the screenshot I showed in the meeting came straight from a single operation of vertex randomization. Nothing else. There's no subdivide happening. I made a short video below to show exactly what I mean.

Blender's documentation doesn't talk about exactly how they do the randomization - that's beyond an artist's interest. We'd have to look at the source code to see how they do it. Judging from the result though, it seems it's a randomized offset of existing vertices. So if you have one vertex offset 0.1 m in one direction, and a consecutive vertex 0.1 m in the opposite direction, then you can get a 180 degree angle.

Let's move the normals discussion to a separate issue, since it's turning out to be a longer thread. You or I can open an issue.

Let's get this PR reviewed and merged first. It's pretty minimal code change, mostly documentation. Do you have the setup to review it? If not, I could re-tag Jessica for the review.

Video of just vertex randomization on a stock cylinder object (radius 1 m, height 1 m), no other modifications:
I show normal = 1.0, then normal = 0.0:

https://user-images.githubusercontent.com/7608908/152463435-9c7d7631-b795-4dab-a53d-bed958c0e7d8.mp4

The ones I showed in the meeting for reference:

Original cylinder
![2022-01-31-191548_3706x2049_scrot_cylinderSubdiv0](https://user-images.githubusercontent.com/7608908/152463495-e1e6bb51-6701-45c4-b0f0-6f20caa2eff3.png)

vertex randomization offset 0.1:
![2022-01-31-194943_3706x2049_scrot_vertRand0 1](https://user-images.githubusercontent.com/7608908/152463545-8e28418d-789d-4bad-9fb9-a1a1bfafb3f1.png)

offset 0.3:
![2022-01-31-194947_3706x2049_scrot](https://user-images.githubusercontent.com/7608908/152463602-1ed8e81b-0363-4977-8351-aa4445698d2a.png)

offset 0.7:
![2022-01-31-194958_3706x2049_scrot](https://user-images.githubusercontent.com/7608908/152463623-925bab19-3a41-4103-b322-765da88f82c5.png)

offset 1 m (same magnitude as cylinder radius and height, 1 m):
![2022-01-31-195015_3706x2049_scrot_vertRand1](https://user-images.githubusercontent.com/7608908/152463642-631c4863-ac6f-40a0-9336-280d8419a33b.png)

j-herman commented:
@mabelzhang I have time this morning - I'll get the review done.  I accidentally updated Blender so I can check if it works on version 3.x and downgrade my computer afterwards for the review... whoops!

mabelzhang commented:
> When running the tutorial as written, blender crashes (both 3.x and 2.92)

Yes! Thank you for retesting and finding that crash. I'm always looking to make my code as robust as possible, but I had not done enough checks in this prototype. Updated in e176ffd

> method variable has to be specified as an array

Updated in README and in code in e176ffd

> path, it shouldn't have a leading '/'

Are you using a relative path? I use absolute path for mine with a leading `/` and it works. I have not tried relative paths. Looks like if I do `os.getcwd()` in the Blender Python prompt, it prints the path of where the Blender executable is. That's pretty far from where DAVE is on my machine, so I've been using absolute paths.

j-herman commented:
Looks good at first review - I'll run it once more when I get back to the other computer and approve.  Now there's no way for me to get it wrong... :)
Yeah, I was using relative path and had just been staring at errors for too long to remember that I didn't need to do that.  Whoops.

woensug-choi commented on urdf/scripts/mesh_distortion/README.md:
```diff
@@ -33,22 +33,30 @@ There are two Coke can models as examples:
 `models/dave_object_models/models/Coke_Can/meshes/coke_can.obj`.
 Choose Transform while importing: Z Up, X Forward.
 
+Tip: To view the textured object in the viewport, click on the globe icon in the
```
Minor suggestion,
We need not only the `obj` file but also `mtl` and  `../materials/coke.png` to load the texture correctly.

mabelzhang replied:
I added a line 5f7030b
Those files are considered part of the object model files, and it's usually assumed that the users have the entire `Coke` or `Coke_can` folder, which would include all the associated files.

Side note: I noticed that you copied some files from `Coke_Can` to a flat directory in the `nps_uw_multibeam_sonar` package, but you should really copy the entire folder, because we use relative paths inside the 3D model files to reference to the `materials`, `scripts`, etc. subfolders. Those folders also contain `model.sdf` and `model.config`, which Gazebo looks for (there's an error about missing `'model.config` for `nps_uw_multibeam_sonar/models/Coke_Can`). All models on Ignition Fuel follow this directory structure.

I would also recommend renaming the copy in `nps_uw_multibeam_sonar`, because if I do catkin install, there are two models of the same name (Coke_Can) in the GAZEBO_MODEL_PATH, and Gazebo picks one of them, such that even when I do `roslaunch dave_demo_launch distorted_coke.launch`, it picks the one in `nps_uw_multibeam_sonar` instead of the one in `dave_object_models` and gives that `model.config` error. Copying that entire folder, and renaming the folder in `nps_uw_multibeam_sonar` should disambiguate between them.

woensug-choi replied:
Gotcha

woensug-choi commented on urdf/scripts/mesh_distortion/README.md:
````diff
@@ -62,15 +70,17 @@ object_prefix = 'coke_can'
+sys.argv = ['distort.py', file_path, object_prefix, distort_extent, method]
 ```
 
 Run the script, replacing its path with the one on your machine:
````
Path description in the distorted mesh's `mtl` file uses `absolute` path. The original `coke.mtl` wrote `map_Kd ../materials/coke.png
`. The distorted file produced by the distort.py script writes `map_Kd C:\\Users\\woens\\Downloads\\MeshDistortion\\Coke\\materials\\coke.png`. I think it would be better to keep to the relative.

mabelzhang replied:
Good catch. I added a `path_mode='RELATIVE'` for the OBJ export, which exposed that in the API https://docs.blender.org/api/current/bpy.ops.export_scene.html#bpy.ops.export_scene.obj 5f7030b

Side note: The COLLADA export (`bpy.ops.wm.collada_export`) though doesn't have that in the API. We aren't using that right now, as it doesn't export textures correctly, so it shouldn't affect the instructions. I added a TODO item above that line.

woensug-choi commented on urdf/scripts/mesh_distortion/README.md:
```diff
@@ -33,22 +33,30 @@ There are two Coke can models as examples:
+Tip: To view the textured object in the viewport, click on the globe icon in the
+upper right of the viewport, which is for Viewport Shading: Material Preview.
+
 Look for the mesh name in the model.
```
Minor suggestion,
I was trying to use custom 'object_prefix' other than 'LPCoke_Cube'. Failed, of course. The naming 'prefix' gives me a feeling that its customizable (got lazy to type 'LPCoke_Cube' and typed 'test' instead). Wouldn't it be better to change `object_prefix='LPCode_Cube'` to `object_name='LPCode_Cube'`?

mabelzhang replied:
There is a specific reason for it to be called `prefix`. If you run the script multiple times, the mesh objects will be imported with the same name, but with incremented suffix `.001`, `.002`, `.003`, etc. (I don't know why Blender does that. It does that if you do it manually too, even if you delete all the meshes in the scene.)

So we cannot just look for an exact match in the object name. We have to look for the prefix. The Python line that looks for the object uses `startswith()`.

woensug-choi replied:
Understood. I should learn `startswith()`

woensug-choi reviewed (changes requested):
Also, I think we should change `fouling_rating` to `distort_extent` at `coke_templtae.erb` 
https://github.com/Field-Robotics-Lab/dave/blob/9eaee7f3bde44841c57511841a920e2a0db76d8e/urdf/scripts/sdf_custom_properties/coke_template.erb#L22
https://github.com/Field-Robotics-Lab/dave/blob/9eaee7f3bde44841c57511841a920e2a0db76d8e/urdf/scripts/sdf_custom_properties/coke_template.erb#L92-L93

Another minor suggestion is add a link to 2.92 blender downloads: https://download.blender.org/release/Blender2.92/

So, at the current state, a distorted mesh obj file is created when the python script for a blender is performed. Then, Ruby erb script is used to make an sdf. Could we link these procedures into one? (this may be another PR or an issue).

I've used python script and template to produce a complete set of a model object by putting all necessary files into a separate model directory with mesh, texture, and sdf when generating dissected bathymetry tiles. I may be able to reconstruct and link two procedures into one using the python console on the blender. (S-size task) This maybe useful when linking the `distort extent` to automatically calculate and write a custom SDF tag for plugins.
If you want to give it a peek about how it's done (uses linux `sed` function with `os.system`), 
- Python code block : https://github.com/Field-Robotics-Lab/Bathymetry_Converter/blob/0eeb120ac6c72d3bd28a6177987222a87a2c120a/mkbathy.py#L130-L150
- Template : https://github.com/Field-Robotics-Lab/Bathymetry_Converter/tree/master/templates

j-herman commented:
@mabelzhang I ran the code again and couldn't break it!  Looks good from the script perspective.  @woensug-choi made a good point about updating the demo erb and the custom_surface_properties_plugin to match the new terminology and ranges here - do you want to add those updates to this PR or are they coming separately?

mabelzhang commented:
I believe 5f7030b addresses the comments.

I debated when I opened the PR whether to rename the fouling parameter in `sdf_custom_properties`, since we weren't entirely sure whether to go the biofouling way for both the mesh modification and the Gazebo side. Now I guess we might as well. It adds a few more files to the diff - the SDF file and the Gazebo plugin.

> add a link to 2.92 blender downloads

I added the link. I also reworded the README to clarify this, because the link looks like a raw folder that users shouldn't go to. They don't necessarily need 2.92. It is not a LTS, so I don't want to recommend that specific version; it's just the version I've tested in. 2.93 is LTS but I haven't tested there. The instructions should work for newer versions. The mentioning of 2.92 is because the Blender UI changed significantly in 2.92, so item locations, shortcuts, etc. as written might not work for older versions. They should work for newer versions though.

> Could we link these procedures into one?

I don't know if that's possible, because the Blender Python script is executed from a Python interpreter within the Blender GUI, whereas the Ruby script is run on the command line, outside Python.

I also don't know whether it's desirable to couple the two scripts. They take different parameters. The mesh distortion script is an artistic process for a general model file, whereas the SDF Ruby template is a physical process specific to parameters that will affect results in Gazebo.

We happen to have the biofouling rating in both, but you can see they weren't the same name (now they are, but) as the two scripts weren't intended to be coupled. Somebody could add any custom SDF tags to their file, like material properties, with or without distortion. The fouling / distort SDF tag is just an example to illustrate the users can add whatever they want. If the user doesn't have any distortion, then there's really no relation between the two scripts to warrant linking them.

Right now, it doesn't take too much effort for us to put in the same distort value for both. I don't know if it'll be worth the effort to link them. I also think that's something we can leave up to the user to do, as opposed to handling in DAVE.

Less relevant, but when we upstream these two scripts, they will live in different repos. One will be in an SDF repo, and the other will be in `ign-common`, precisely because they do such separate things.

woensug-choi commented:
FYI, @mabelzhang I've updated the multibeam sonar repo for the duplicate model directory. It's not in SDF file structure but a bare OBJ and MTL files now in the `coke_can_example` directory. I think it's better for the sonar tutorial perspective that all SDF custom tags can be seen in a single `world` file to play with.
